### ISSUE TITLE
Fuzzy completion woes

### ISSUE THREAD

**Issue by alexander-yakushev** [clojure-emacs/cider#1909]
So, I've been using company together with company-quickhelp for a while now. For the most part, I'm pretty content with how everything works, except for fuzzy completion. I've tried the `company-flx` route as described in the docs, but it doesn't solve the problem:

- `company-flx` works by polling the backend with empty prefix (`""`) and then performing the filtration on the client-side. This does a poor job with Compliment since the latter doesn't spit out _absolutely all_ candidates when queried with `""`. E.g. a regular project has tens of thousands of classes loaded, and we can't just return all of them. Thus, Compliment actually expects at least something in the query to start looking into heavyweight sources like classes and methods.
- `company-flx` is *too fuzzy*, using very wide regexp to filter candidates. Even despite the intelligent ranking algorithm, I am quite confused by some of the suggestions it offers. This may be subjective, of course.

I could let some of the missing fuzziness go, if not for [simple classname completion](http://bytopia.org/2015/03/06/smart-classname-completion-for/). I do plenty of interop these days, and it is absolutely painful having to Google that to import `ProgressEvent` I actually have to write `com.amazonaws.services.s3.model.ProgressEvent`.

Long story short, I've made a blunt hack _for myself_.

```elisp
(defun cider-company-capf-advice (old-fun &rest args)
  (let ((completion-styles (list 'cider)))
    (apply old-fun args)))

(defun cider-company-completions (string table predicate point)
  (cider-complete string))

(advice-add 'company-capf :around #'cider-company-capf-advice)
(add-to-list 'completion-styles-alist
             '(cider
               cider-company-completions
               cider-company-completions
               "CIDER backend-driven completion style."))
```

Basically, I add an adivce to `company-capf` that forces it to skip all its machinery and call straight to `cider-complete`. This works for me because I don't use company for anything except CIDER — but I can imagine a solution like this breaks all other company backends.

I create this issue with the intent to discuss the following points:

- Is fuzzy completion of any value? I vote yes, especially for the classname example above. That's certainly Cursive's strong suite.
- Should we continue promoting `company-flx`? I don't feel like it offers good UX for CIDER particularly.
- Can the advice hack be improved and embedded into CIDER by default? From what I see we don't need more than UI from company/standard Emacs completion mechanisms, so should we limit our functionality because of aribitrary restrictions in their "logic" part?

**Comment by expez** [comment]
@dgutov, do you have any suggestions?  I doubt @alexander-yakushev is the only person missing fuzzy completions.  

Since dealing with fuzzyiness from emacs appears to be bad from a performance standoint, perhaps we can send a message like `{:prefix "foo" :fuzzy true}` to the backend and only rely on company to display the results?

**Comment by bbatsov** [comment]
Guess you've seen this ticket https://github.com/company-mode/company-mode/issues/47, right?

**Comment by alexander-yakushev** [comment]
@bbatsov Sure did. Dmitriy reasonably says that company is limited by Emacs' own completion mechanism (`all-completions` function as I discovered). It would be difficult and unwise for him to start reinventing it. `company-flx` is then claimed to be the solution, but it works poorly for us for the reasons I outlined above.

Once again, we don't really need neither company, nor company-flx, to deal with our completion. We can just use company as UI.

@expez I don't really want to drag Dmitriy into this once again, I think our usecase is too specific too try to integrate into the bigger picture. Unless Dmitriy can suggest some good way to skip the company machinery and not rely on a hack.

> perhaps we can send a message like {:prefix "foo" :fuzzy true} to the backend and only rely on company to display the results?
That's what I'm talking about. Compliment already returns fuzzy candidates by default, so we only need the Emacs-side part to stay out of the way and not filter the returned candidates further.

**Comment by bbatsov** [comment]
> That's what I'm talking about. Compliment already returns fuzzy candidates by default, so we only need the Emacs-side part to stay out of the way and not filter the returned candidates further.

I we can just add a simple company-backend that would take precedence over company-capf. Users will have to enable it manually, but that's not a big deal.

**Comment by alexander-yakushev** [comment]
Sounds good, looking into implementing custom backends now. Is there a reason this was not done from the beginning? Any drawbacks?

**Comment by bbatsov** [comment]
> Sounds good, looking into implementing custom backends now. Is there a reason this was not done from the beginning? Any drawbacks?

Nope. Just wanted to rely on the standard `capf` behaviour instead of introducing some custom solution for `company`. Haven't used an IDE for over 6 years (when I was a Java dev) and I've forgotten about the usefulness of real fuzzy completion. :-)

I'm guessing implementing this should be trivial, so it's better to add it than to inflict pain on the users who need it.

**Comment by dgutov** [comment]
Like others, I don't think using `company-flx` is ideal for backends that perform fuzzy matching themselves. This (fuzzy matching in the backend) seems a common/desired enough case, though.

Last time we discussed that with @monnier, we've settled on the understanding that we need to standardize a new kind of "completion table" (or extend the current definition somehow) that would allow to delegate all the heavy lifting in the fuzzy matching to the background process. And teach the current completion UI about it. As an aside, asynchronous invocation was another feature-to-add.

I haven't got around to it thus far, and it seems nobody else is invested enough, unfortunately.

**Comment by dgutov** [comment]
But!

@alexander-yakushev Does your workaround function well enough for CIDER completion? If yes, why not, instead of advice, make a buffer-local setting for `completion-styles` in `cider-mode`? And modify `completion-styles-alist` to include the `cider` entry, either locally or globally; either way should do.

That would mess with other completion functions in the CIDER-enabled buffers, but the linked pull request recommends the users to leave only `company-cider` in `company-backends` anyway, so working with other backends poorly doesn't seem to be considered a problem. And this would only affect other `completion-at-point-functions` anyway.

**Comment by bbatsov** [comment]
> That would mess with other completion functions in the CIDER-enabled buffers, but the linked pull request recommends the users to leave only company-cider in company-backends anyway, so working with other backends poorly doesn't seem to be considered a problem. And this would only affect other completion-at-point-functions anyway.

Well, the linked PR doesn't interfere with other backends - it really interferes with `company-capf` I guess.

**Comment by alexander-yakushev** [comment]
Thanks @dgutov!

I don't really understand the separation between company "backends" (controlled by `company-backends`) and company "frontends" (controlled by `completion-styles` AFAIU). I used the former in the related PR, and it seems to be working fine. Is there a benefit to use `completion-styles` instead?

**Comment by dgutov** [comment]
> Well, the linked PR doesn't interfere with other backends - it really interferes with company-capf I guess.

Oh, right. But the effect of changing `completion-styles` would also be limited to `company-capf`.

> and company "frontends" (controlled by completion-styles AFAIU)

Nah, `completion-styles` is not a part of company. It's a variable in `minibuffer.el` that decides how `completion-at-point` performs matching. `company-flx` also uses it to install its fuzzy matching middleware.

`completion-styles` affect the results of `company-capf`.

> Is there a benefit to use completion-styles instead?

* The users don't need to do any extra setup.

* Fuzzy completion might work with `C-M-i`, for non-company users. (Does it? Someone should test that).

**Comment by alexander-yakushev** [comment]
So, it's the approach I used from the very beginning. Although, it involves adding an advice around `company-capf`. Is there a better way to set `completion-styles`? I also have a feeling that @bbatsov won't be happy modifying core functionality automatically, so in the end it might be something that a user would have to do manually anyway.

**Comment by dgutov** [comment]
> Is there a better way to set completion-styles?

With `setq-local`, somewhere near to where you modify `completion-at-point-functions`.

**Comment by alexander-yakushev** [comment]
What do you say, @bbatsov? Can we enable the fuzzy automatically? Should we make it configurable and enabled by default? Configurable and disabled by default?

**Comment by bbatsov** [comment]
Configurable and disabled by default. This is a major change.

**Comment by alexander-yakushev** [comment]
Given that, until the last year, most people used `auto-complete` and had this behavior by default, this is not a major change to many. But fine, I will do that.

**Comment by bbatsov** [comment]
Ever since I started working on the project I've been promoting the use of `company-mode`. Not to mention the old completion mechanism of CIDER didn't have a particularly advanced fuzzy matching support. :-) I'm an Emacs user first and a Clojure user second, so I generally prefer default settings to be aligned with Emacs common conventions.

**Comment by alexander-yakushev** [comment]
On a second thought, I don't see the necessity to break into `capf` if we make the feature opt-in. The implementation can remain the current one, with a custom backend.

Making it "configurable via a vairable" also requires more effort than "add these lines to the config".

**Comment by dgutov** [comment]
That's not "breaking in". Changing `completion-styles` is a legitimate approach.

But someone should test whether `C-M-i` continues to work fine (or better).

**Comment by bbatsov** [comment]
Let's go with the custom backend. It seems like the more straightforward approach to me.

**Comment by bbatsov** [comment]
Fixed by #1910.
